Compositing notebooks and functions

cbur24 opened the pull request:
### Proposed changes
Started redoing the `xr_geomedian` notebook and then decided to port over from DE Africa all things image compositing.  Two new notebooks: `Generating geomedian composites` and `Generating composites`; five new functions for `dea_datahandling`: `calc_geomedian`, `first`, `last`, `nearest`, `_select_along_axis`

I haven't run these on the NCI.

### Checklist (replace `[ ]` with `[x]` to check off)
- [X] Notebook created using the [DEA-notebooks template](https://github.com/GeoscienceAustralia/dea-notebooks/tree/develop)
- [X] Remove any unused Python packages from `Load packages`
- [X] Remove any unused/empty code cells
- [X] Remove any guidance cells (e.g. `General advice`)
- [X] Ensure that all code cells follow the [PEP8 standard](https://www.python.org/dev/peps/pep-0008/) for code. The `jupyterlab_code_formatter` tool can be used to format code cells to a consistent style: select each code cell, then click `Edit` and then one of the `Apply X Formatter` options (`YAPF` or `Black` are recommended).
- [X] Include relevant tags in the final notebook cell (refer to the [DEA Tags Index](https://docs.dea.ga.gov.au/genindex.html), and re-use tags if possible)
- [X] Clear all outputs, run notebook from start to finish, and save the notebook in the state where all cells have been sequentially evaluated
- [ ] Test notebook on both the `NCI` and `DEA Sandbox` (flag if not working as part of PR and ask for help to solve if needed)
- [ ] If applicable, update the `Notebook currently compatible with the NCI|DEA Sandbox environment only` line below the notebook title to reflect the environments the notebook is compatible with

pep8speaks commented:
Hello @cbur24! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`Scripts/dea_datahandling.py`](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py):

> [Line 718:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L718): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 718:23](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L718): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 719:32](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L719): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 720:35](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L720): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 722:34](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L722): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 723:29](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L723): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 727:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L727): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 728:69](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L728): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 730:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L730): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 737:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L737): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 738:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L738): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 747:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L747): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 749:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L749): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 758:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L758): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 761:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L761): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 764:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L764): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 770:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L770): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 773:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L773): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 779:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L779): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 780:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L780): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 785:78](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L785): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 786:79](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L786): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 792:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L792): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 798:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L798): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 801:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L801): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 811:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L811): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 816:78](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L816): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 817:79](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L817): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 824:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L824): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 830:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L830): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 834:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L834): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 836:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L836): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 838:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L838): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 839:78](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L839): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 840:72](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L840): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 842:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L842): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 845:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L845): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 857:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L857): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 862:78](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L862): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 868:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L868): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 871:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L871): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 872:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L872): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 873:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L873): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 874:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L874): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 879:1](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L879): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 881:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L881): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 883:80](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L883): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 885:63](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L885): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 886:68](https://github.com/GeoscienceAustralia/dea-notebooks/blob/442f014f5f6d07f0713645385f58cf7c7e826413/Scripts/dea_datahandling.py#L886): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

robbibt commented:
Reviewing now!

Kirill888 commented on Scripts/dea_datahandling.py:
```diff
@@ -704,3 +714,202 @@ def paths_to_datetimeindex(paths, string_slice=(0, 10)):
+    da = (odc.algo.reshape_for_geomedian(ds, axis=axis) if
+          isinstance(ds, xr.Dataset) else ds)
+
+    if max_value is None:
+        max_value = da.max(skipna=True)
+    if min_value is None:
+        min_value = da.min(skipna=True)
```
So this will be really slow for dask inputs, it will essentially load all the data three times, one for max one for min and then again for actual computation.

Kirill888 replied:
also it is likely to be broken for `nodata=-999` case as it will pick `-999` as min values, as I do not think that `skipna` skips `nodata` values, only float NaNs

Kirill888 replied:
Scaling is only performed to improve numeric stability it is not a prerequisite for correct operation, so I would just skip scaling and convert to float with `*1+0` if user hasn't supplied min/max.

Also from API perspective maybe easier if that was something like `value_range=(min_val, max_val)` or some better name.

cbur24 replied:
For DE Africa, its handy having the `max_value` and `min_value` parameters as the satellite products have different surface reflectance scalings. Are the SR scaling values identical between all products for DEA? I think they might be (0-10,000), in which case we could remove the calculations and hard code the min and max values.

Kirill888 replied:
also checkout `int_geomedian`, I think that gives better performance as it delays conversion to float until very end, just before calling to Dale's code, so it has lower memory and bandwidth requirement.

Kirill888 replied:
Have you observed significant difference of output quality when selecting "exact" scaling factors vs none? 3 times slower operation and 3 times bandwidth better be worth it.

Kirill888 replied:
Besides we are not preventing that use case, just changing default behaviour to be less "surprising". We can just document the fact that scaling can help somewhat and if you know rough range of the data you can supply it and maybe get slightly better results out of it. And if not -- nothing bad.

With code as is, it is broken when given completely uniform input, as is will attempt `/0`

Kirill888 commented on Scripts/dea_datahandling.py:
```diff
@@ -704,3 +714,202 @@ def paths_to_datetimeindex(paths, string_slice=(0, 10)):
+    max_value : the maximum value that could be found in the array
+    min_value : the minimum value that could be found in the array
+    '''
+    da = (odc.algo.reshape_for_geomedian(ds, axis=axis) if
+          isinstance(ds, xr.Dataset) else ds)
```
`xr_geomedian` does accept `xr.Dataset` objects now and performs reshape internally.

cbur24 replied:
Thanks @Kirill888, can you update the code in this PR to reflect this change? And then I will port it across to DE Africa.

Kirill888 replied:
I think `odc.algo.int_geomedian` is exactly this function, but has different internal implementation that takes advantage of knowing that input is `int`.

`int_geomedian` does assume `time,y,x` dimension order, but all of our data is like that, and instead of min/max it takes scale offset with default being no scale no offset

Kirill888 replied:
hold on, in the example notebook loaded data is already `float32` cause this happened inside `load_ard`

robbibt replied:
Ah yes, if the data was loaded in `load_ard` with cloud masking selected it will be in `float32`

robbibt replied:
I think we should be able to get it back as `Int` if we add `dtype='native'` to the function call

Kirill888 replied:
I reckon staying in `int16` throughout is preferable.
`xr_geomedian` for float inputs, `int_geomedian` for int16 inputs, ideally I should merge it into one function though.

robbibt replied:
Having them as one function would be really nice, `xr_geomedian` is a nice name for a general purpose wrapper function I think

robbibt commented:
Hey @Kirill888 and @cbur24 , I currently have the branch checked out and am making changes, so I'm happy to refactor the code RE Kirill's suggestions above if that works. Just so I have it right: we can probably replace the entire `calc_geomedian` func with `odc.algo.int_geomedian` instead?

cbur24 commented:
@robbibt if you could incorporate @Kirill888 suggestions that would great.  It seems like we probably can remove the `advanced control for geomedian` part.  Maybe we could instead show the use of `int_geomedian` and `xr_geomedian`? With a few words about how they differ? Something about speed and memory vs the desired output?

Kirill888 commented:
yep what @cbur24 said above.

Needs a note about scaling factors: basically aim for roughly 0-1 range after scaling, but it doesn't need to be precise and can always default to operating on original values.

robbibt commented:
Hey @cbur24 and @Kirill888 , I've made a first pass at re-writing the notebook. I've changed up the order a fair bit for a simpler/more logical flow (load data > int_geomedian > grouping and resampling). `int_geomedian` works nicely after changing `load_ard` to `dtype='native'`, so I've added a bit up the top about memory use and why keeping the data in Int is good.

That said, I don't think I have enough knowledge about the workflow required to use `xr_geomedian` to write about it in the notebook. I don't know what parts of the "advanced" section are required as a bare minimum example, so I've had to leave it out for now. I think if either of you wanted to cover it, putting it in as a final section would be good so users aren't scared away by scaling factors etc earlier on.

robbibt commented:
I've also had a quick run through of the non-geomedian composite notebook and made some minor changes. I really like it, but there's something very strange going on in one of the plots:

![image](https://user-images.githubusercontent.com/17680388/87750858-e80ecb80-c83f-11ea-9dc0-a9e9174bbca5.png)

That looks to me like possibly something confused with axes being plotted in the wrong order... @andrewdhicks?

Kirill888 commented:
@robbibt thanks, this looks good. `xr_geomedian` is basically the same as `int_geomedian` except it assumes that input data is already in `float32` format. It can also operate directly on `xr.DataArray` and even on numpy arrays. But most should interface with it using `xr.Dataset`. 

When input is a single array the dimensions ought to be in this order `y, x, band, time`, with `time` dimension being removed by the process. This is awkward order, but that's what Dale's code expects natively, I guess it optimizes for computation performance.

I think the whole scale/offset can be ignored in the notebook, I do need to document it better in the function doc comment though.

cbur24 commented:
@Kirill888 I've added a section at the end of the geomedian notebook demonstrating the use of `xr_geomedian`, if you could please give it a quick look and see that it makes sensethat'd be great.  Overall, I think the notebook is looking really good now. I've also removed the `calc_geomedian` code from `dea_datahandling`.

@robbibt I think I resolved the 'age of pixels' plot. The issue was with this line:
```
# Identify last time in our data
last_time = last_blue.time[-1]
```
Identifying the 'last' time isn't what we wanted, we wanted the _latest_ time, because then we can compare each pixel in the 'last_blue' composite with the lastest time.  ie
```
# Identify lastest time in our data
last_time = last_blue.time.max()
```

Kirill888 commented:
actually I default `num_threads=1` in my code, so doesn't matter what Dale's defaults are.

cbur24 commented:
Thanks @Kirill888.  I placed `num_threads=1` in there because without it the `xr_geomedian` code printed "working on 16 threads", wasn't sure if that was just an errant print statement or if it was having an impact, either way it was confusing so I think its fine to leave in there for now.

Kirill888 commented:
looks like it's only `int_geomedian` that sets the default value to `1`, I should change `xr` version to do the same.

robbibt commented:
@cbur24 Just a small typo in the composite notebook: 
![image](https://user-images.githubusercontent.com/17680388/88236798-30efd580-ccc1-11ea-9fb6-a2f170069936.png)

Any way we can remove the FutureWarning by adding the param they recommend?
